### ISSUE TITLE
Cannot open dialog using webpack bundle

### ISSUE THREAD

**Issue by mib4fun** [aurelia/dialog#127]
Hello,

I'm having trouble making work aurelia-dialog withwebpack.
Webpack is loading correctly the plugin. Here is my sample repo : 
https://github.com/mib4fun/aurelia-dialog-sample

You have to delete the require css line  in the commonjs dist to make it work.
I have described it here : https://github.com/mib4fun/aurelia-dialog-sample#getting-started 

Then when i try to open a dialog i get this error
 
![error](https://cloud.githubusercontent.com/assets/3010867/14046160/4d5bc66a-f297-11e5-8265-0dc60ab7ab5f.jpg)

**Comment by miguelzakharia** [comment]
@mib4fun Did you by any chance have these errors before getting as far as you did? I can't even get my page to load because it can't find references.

![image](https://cloud.githubusercontent.com/assets/682476/14062252/7b94ffc4-f36e-11e5-8daf-bd68b66ae431.png)

**Comment by mib4fun** [comment]
@miguelzakharia
I have updated the repo with my fork of the aurelia-dialog.

It loads fine on heroku here is the link : http://aurelia-dialog-sample.herokuapp.com/
( don't mind the sockjs errors )

**Comment by kuechlerm** [comment]
@mib4fun even with your fork, I still get this error when trying to open a dialog:

`
Uncaught (in promise) TypeError: Cannot read property 'canActivate' of undefined
    at invokeLifecycle (webpack:///./~/aurelia-dialog/dist/commonjs/lifecycle.js?:8:22)
    at eval (webpack:///./~/aurelia-dialog/dist/commonjs/dialog-service.js?:56:47)invokeLifecycle @ lifecycle.js:8(anonymous function) @ dialog-service.js:56Promise.resolve (async)(anonymous function) @ dialog-service.js:53open @ dialog-service.js:41Editor.canDeactivate @ editor.ts:74iterate @ aurelia-router.js:1082processDeactivatable @ aurelia-router.js:1092run @ aurelia-router.js:1024next @ aurelia-router.js:127(anonymous function) @ aurelia-router.js:535Promise.resolve (async)_buildNavigationPlan @ aurelia-router.js:605run @ aurelia-router.js:533next @ aurelia-router.js:127run @ aurelia-router.js:140(anonymous function) @ aurelia-router.js:1562Promise.resolve (async)_dequeueInstruction @ aurelia-router.js:1535(anonymous function) @ aurelia-router.js:1526_queueInstruction @ aurelia-router.js:1523(anonymous function) @ aurelia-router.js:1457Promise.resolve (async)loadUrl @ aurelia-router.js:1456_loadUrl @ aurelia-history-browser.js:279_checkUrl @ aurelia-history-browser.js:272
`

Any solutions / workarounds?

**Comment by seesharper** [comment]
I have the same problem as @kuechlerm

**Comment by miguelzakharia** [comment]
Same here. I tried debugging, but no luck.

**Comment by seesharper** [comment]
Try referencing the viewmodel by module name

``` 

submit() {
    this.previousValue = this.fullName;
    this.dialogService.open({
      viewModel: 'prompt',
      model: `Are you ${this.fullName} ?`
    })
    .catch((err) => {
      console.log(err);
    })
  }

```

**Comment by mib4fun** [comment]
@seesharper thx ! It worked !!

**Comment by jww2000** [comment]
@seesharper & @mib4fun & @swamipras I got a little further now I get DialogRenderer must implement getDialogContainer()

````code
Unhandled rejection Error: DialogRenderer must implement getDialogContainer().
    at Renderer.getDialogContainer (webpack:///./~/aurelia-dialog/dist/commonjs/renderers/renderer.js?:15:11)
    at eval (webpack:///./~/aurelia-dialog/dist/commonjs/dialog-service.js?:61:86)
From previous event:
    at eval (webpack:///./~/aurelia-dialog/dist/commonjs/dialog-service.js?:56:82)
From previous event:
    at eval (webpack:///./~/aurelia-dialog/dist/commonjs/dialog-service.js?:51:111)
From previous event:
    at eval (webpack:///./~/aurelia-dialog/dist/commonjs/dialog-service.js?:49:47)
From previous event:
    at DialogService.open (webpack:///./~/aurelia-dialog/dist/commonjs/dialog-service.js?:39:19)
    at Welcome.getDialog (webpack:///./src/welcome.js?:60:24)
    at CallScope.evaluate (webpack:///./~/aurelia-binding/dist/commonjs/aurelia-binding.js?:1453:19)
    at Listener.callSource (webpack:///./~/aurelia-binding/dist/commonjs/aurelia-binding.js?:4882:40)
    at eval (webpack:///./~/aurelia-binding/dist/commonjs/aurelia-binding.js?:4906:22)
    at HTMLDocument.handleDelegatedEvent (webpack:///./~/aurelia-binding/dist/commonjs/aurelia-binding.js?:3115:5)
````

**Comment by mib4fun** [comment]
@jww2000 can you compare that with my sample   https://github.com/mib4fun/aurelia-dialog-sample
It's working all good now.

**Comment by jww2000** [comment]
@mib4fun awesome I was missing the plugin in main.js! thank you so much! It is finally working after 2 weeks ;)

**Comment by mib4fun** [comment]
Cool ! i close the issue now

**Comment by seesharper** [comment]
Why are you closing this? The original issue is not actually resolved. It should still be possible to reference the view model by its type as you can when running JSPM. I think it is important to leave this open so that it can be fixed by the Aurelia team.

**Comment by mib4fun** [comment]
woups... you right, reoping.
@PWKad, can you take a look please ?

**Comment by EisenbergEffect** [comment]
Probably it's missing functionality in the webpack module loader for
Aurelia.

**Comment by cdennig** [comment]
Assigning the viewmodel as a string only works, when it is in the same directory. When you want to reference one that's not in the same folder, you'll get an error: "Cannot find module './DialogViewModel'."
Please have a look at the issue, because there is no workaround at the moment. Thanks in advance...

**Comment by Tseberechts** [comment]
@cdennig you can say 'folder/DialogViewmodel' (path from src to model'

**Comment by The1nternet** [comment]
The original issue states "You have to delete the require css line in the commonjs dist to make it work."  This is still the case, as I just had to delete that same line to make aurelia-dialog work with webpack.

**Comment by PWKad** [comment]
@EisenbergEffect We can leave this open here for now but sounds like some functionality is missing in the webpack loader that is needed to fix it - I'll open an issue there.

**Comment by Vidarls** [comment]
Just want to summarize all the issues I experienced today trying to use aurelia-dialog using typescript and webpack. They might all be covered by different issues on different projects but I feel a summary of dev pain might be useful:

Note: this is all non-happy-path issues from starting with latest skeleton to getting a working modal dialog:

1. `typings` not installed, `npm install -g typings` should probably be part of the README of skeleton.
2. Some core aurelia modules was missing from the skeleton typings file (aurelia/skeleton-navigation#401). Note: it is generally a pain to manually add stuff to the typings config file when installing modules.)
3. Had to manually add `aurelia-dialog` to typings file.
4. Had to manually fix webpack.config  iaw @mib4fun gist to have webpack properly load modules. Undocumented, and one more manual step just to add a module
5. Had to remove the css require in `node_modules/aurelia-dialog/dist/commonjs/resources/ai-dialog.html` will of course break in any CI build, or when the module is updated.
6. Had to manually re-add the css file to get the correct modal looks.
7. Had to refer to module by string name rather than Type in `DialogService.Open(...)` to make it work. Undocumented and not intellisense friendly.

This might look like a rant, but it is not. Promise.. 

Just wanted to summarize the issues while fresh in my mind, to help make Aurelia the most developer friendly framework possible.

**Comment by Vidarls** [comment]
Small update, tried the same flow using the JSPM typescript skeleton.

1. The JSPM typescript does not require you to run `typings install` hence problem no 1 through 3 above was not present
2. No manual changes was required for adding the `aurelia-dialog` plugin in any files (except of course adding `.plugin('aurelia-dialog')` to the bootstrapper.
3. No issues with CSS loading
4. For some reason even problem 7, requiring to refer to viewModel by string name was not an issue, just using the imported class directly (as normal) was working.
5. Needed to run an additional JSPM update after adding `aurelia-dialog` to resolve some JSPM created forks.

tl;dr;

JSPM + typescript behaves as expected minus some fork issues with JSPM.

**Comment by EisenbergEffect** [comment]
@Vidarls Thanks for putting the time in to identify the issues! I do think some of this is related to our core webpack support that we are still working on improving. There's been some discussion related to some of these issues elsewhere. I think we need a webpack guru to help us find the proper solutions, most likely. If you or anyone you know has more experience with webpack plugins and loaders, please put them in touch.

**Comment by activebiz** [comment]
+1 @cdennig ... same issue here. It needs to be addressed asap. Problem is with aurelia-metadata is not loading the viewmodel when its an object. It works when you load it viewmodel name as string because its passing to aurelia-templating engine. To me this is not a fix. It should be able to load viewmodel as object.

**Comment by EisenbergEffect** [comment]
Unfortunately, I don't think it's going to be possible with Webpack. Webpack isn't really a module loader and thus there is no module registry. Without the registry, we cannot lookup the id for a type and thereby load its view. You can see the code for the loader abstraction for webpack here: https://github.com/aurelia/loader-webpack/blob/master/src/index.js  Most module loaders have a registry which we can search. We try to fake this for Webpack since it doesn't have that. However, it won't work 100% of the time. If you can find a way to fix this, we are open, but it's likely a limitation of Webpack itself. As a workaround, try adding a `useView('./path/to/view')` decorator on your modal dialog class. If you do that, it shouldn't need to look up the module id to conventionally load the view. If *that* fails, that is a bug we can probably fix in Aurelia itself. Let me know.

**Comment by activebiz** [comment]
Thanks @EisenbergEffect . 

+1 for aurelia-cli

**Comment by activebiz** [comment]
Have tried the useView() but I still get the same error 

`uncaught (in promise) TypeError: Cannot read property 'canActivate' of undefined(…)`

**Comment by jedd-ahyoung** [comment]
This may or may not be a legitimate bug. The dialog manually attempts to
get the viewModel with a helper method in dialog-service that uses
`Origin.get()` or `compositionEngine.ensureViewModel()` - this may not be
honoring the useView decorator. We may be missing a higher-level
abstraction that we could use for this, but I'm not sure. I'd have to look
into it further.

2016-04-29 1:48 GMT-04:00 Preyash Desai <notifications@github.com>:

**Comment by activebiz** [comment]
okay. I think I missed this earlier. But you can use the bootstrap code in main.ts to load the dialog as globalresource like .

```
    aurelia.use
      .standardConfiguration()
      .plugin("aurelia-dialog")
      .globalResources("common/dialogs/myDialog");
```

Now the issue is that because the ai-dialog.html has 

```
<template>
**  <require from="../dialog.css"></require>**
  <content select="ai-dialog-header"></content>
  <content select="ai-dialog-body"></content>
  <content select="ai-dialog-footer"></content>
</template>

```
it cant load the dialog.css.

**Comment by EisenbergEffect** [comment]
Right. That is the true issue. I think we have another issue somewhere about this. We need to find out how we can tell Webpack to include the dialog.css in the bundle as a text resource. That's the final piece of the puzzle. We need someone who knows Webpack pretty well to help us with this. We'll want to do this with any plugin. For example, if you use plugin 'foo' we want to include all its js, html and css. The html and css needs to be included in the bundle correctly as text.

**Comment by hoangcuongvn** [comment]
Currently, I see the only workaround is assigning to viewModel the dialog's moduleId as an absolute path. Is there somehow to use a relative path instead?

**Comment by valichek** [comment]
Here is the solution how to avoid errors when using `<require from="dialog.css">`. Just two line in webpack config needed.
```
{ test: /aurelia-dialog\/dist\/commonjs\/dialog\.css/, loader: 'raw'},
{ test: /\.css?$/, exclude: /aurelia-dialog\/dist\/commonjs\/dialog\.css/, loader: 'style!css' },
```
First we load it with raw-loader and exclude it from loading with css-loader, and you go

**Comment by EisenbergEffect** [comment]
Awesome. @jdanyow Take note.
@valichek Would you be willing to create a PR to the dialog readme to add instructions for webpack?

**Comment by valichek** [comment]
@EisenbergEffect sure, will create PR

**Comment by mttmccb** [comment]
@valichek I've tried following the instructions on getting webpack working, it works fine in dev but not prod... I've got an example repo here https://github.com/mttmccb/aurelia-webpack-playground

It's basically the skeleton navigation with a few plugins added in.

Any ideas why it might not be working?

**Comment by valichek** [comment]
@mttmccb I have Gist https://gist.github.com/valichek/5dc7c1e11ed39a82df83913df465cb50 with my working configs. Check it, please. It works like a charm for prod config too.

**Comment by mttmccb** [comment]
@valichek thanks, but still getting the same error loading aurelia-dialog/dialog.css

**Comment by bsrdjan** [comment]
Why not simply remove requiring `dialog.css` from `ai-dialog.html` and let the users include the `dialog.css` together with other application styles ? Users (the most of us I think) may need to adapt the default `dialog.css` styling to some overall application styling and if `dialog.css` is required in `ai-dialog.html` that is not possible. I am using compass currently and had to comment the `require` in `ai-dialog.html`, add the `dialog.css` to other compass styles and partly overwrite there.

**Comment by mttmccb** [comment]
Using the raw loader worked, I have no interest in adding less to my project... hopefully there will be simpler options for loading CSS.

**Comment by mattgaspar** [comment]
The RegEx doesn't work on Windows because of the slash direction.
This should work for both Windows and OSX/Linux:

```
{ test: /aurelia-dialog(\\|\/)dist(\\|\/)commonjs(\\|\/)dialog\.css/, loader: 'raw' },
{ test: /\.css$/, exclude: /aurelia-dialog(\\|\/)dist(\\|\/)commonjs(\\|\/)dialog\.css/, loader: 'style!css' },
```

**Comment by hoangcuongvn** [comment]
Suppose the app folder is like below:
src
|__my-feature
....|__my-module.js
....|__my-module.html
....|__my-dialog.js
....|__my-dialog.html

Inside the MyModule vm class, if we have a relative moduleId string (e.g "./my-dialog") we can calculate its absolute moduleId string (e.g "my-feature/my-dialog"). The code is something like below:
```
let o = Origin.get(this.constructor);
let dialogModuleId = relativeToFile("./my-dialog", o.moduleId);
this.dialogService.open({
  viewModel: dialogModuleId,
  model: something,
});
```

Is it possible if aurelia-dialog adds a property i.e "target" into the settings? So we can use relative paths when opening dialog.
```
this.dialogService.open({
  target: this,
  viewModel: "./my-dialog",
  model: something,
});
```

**Comment by sapiens** [comment]
I have a similar issue, but not with css as I don't use it.  In my case, I get an `instance is undefined` TS error. This happens only when the view model is in the current dir and it's imported as a relative path `./viewModel` . The workaround for now is to specify the name of the view model as a string representing the module path relative to the app src dir, like this `common/stuff/viewModel`

**Comment by PWKad** [comment]
The initial problem here with css not loading should be fixed because the css is now loaded from within the view-model directly.  I'm going to close this out for now as the original issue was with the CSS not loading. There is a separate issue that is currently open about referencing module paths with webpack tracked here #213 if anyone thinks that they are different I'm willing to open this back up but it seems to be a webpack loader issue, not so much a dialog issue.
